### PR TITLE
Initial sketch of the Dataset portion of the Search API

### PR DESCRIPTION
**Pull request by jfuerth**
This is an early sketch to try and make the discussion around Datasets concrete. Known loose ends include:

- expected error cases (eg. spec 404 for datasets that don’t exist?)
- how to define units (eg. years, kilograms, centimetres, …)
- add pagination for /datasets list resource
- should we use the word ‘schema’ to describe the structure of a dataset? (aligns nicely with JSON Schema, but is overloaded with the typical meaning of ‘schema’ in the relational database sense)
- discussion about how authentication should work (spec could suggest OAuth access tokens? This would dovetail nicely with work that we're aware of/involved in in the Researcher Identity workstream)

Credit: This is derived from earlier API design work done by Michal Linhard at DNAstack.

### CODE REVIEW

**Comment by mcupak** [comment]
#4

File: DATASET.md
Comment (mcupak) [inline DATASET.md:1]: It would be good to add a link to the MD files in README.md.

File: spec/search-api.yaml
Comment (mcupak) [inline spec/search-api.yaml]: Copy-paste error?

File: spec/search-api.yaml
Comment (mcupak) [inline spec/search-api.yaml]: Would it make sense to start with 3.0.2? I'm not sure what the patches are, just thinking it's probably easier to go with the latest right from the beginning.

File: spec/search-api.yaml
Comment (mcupak) [inline spec/search-api.yaml:7]: This is OK for now, but I filed #5 to address later.

**Review by mcupak — CHANGES_REQUESTED** [review]

**Comment by mcupak** [comment]
@Relequestual I think it would be valuable to have a review from you if you have time, given your expertise in JSON schema and GA4GH search/schemablocks things.

**Comment by jfuerth** [comment]
There is a partial (but functioning) implementation of this proposal that exposes Google Sheets as datasets: https://github.com/DNAstack/ga4gh-search-adapter-google-sheets

**Comment by antbro** [comment]
So do we assume / require
- that units will always be specified in a computer readable way in the description?
- that the dataset is fully denormalised, so that all frequently used values for an attribute are repeated for each object?
- that multiple values for an attributed can be included?
- that multiple values that need to have some order or precedence will capture that in their representation?
- that values for attributes of values for attributes (ad nauseam) will be captured within each object [e.g., amount for sample-volume, for [ID of aliquot-ID and date of date-of-sampling], for ID for sample-ID]?

**Comment by davidlougheed** [comment]
Is there going to be a way to link datasets together under some kind of common identifier/grouping system for joint-querying purposes, or is this out of scope for the search API / is the suggested way to handle this going to be to combine different types of data (ex. variants + expression levels) into a single compound dataset, at least at the API response level? [ex. querying a combined `/datasets/variants+expression_levels`]

**Comment by jfuerth** [comment]
Thanks for taking a look, @antbro!

I'll try to respond to your questions one-by-one.

> * that units will always be specified in a computer readable way in the description?

This PR doesn't mention units, but specifying units for scientific data is certainly essential. In the distant past, I worked with [GNU Units](https://www.gnu.org/software/units/), which has a pretty comprehensive library of base units and conversions based on dimensional analysis (eg. it understands that Newtons are kg*m/s^2). The GPL probably means we won't be able to use GNU Units specifically, but perhaps there is something similar out there with a more compatible license.

I should mention two overriding goals of this spec that probably should be highlighted in the readme:

1. it should be small and easy to implement (give the impression that it would be an enjoyable afternoon project to get a new implementation off the ground in your favourite language)
2. it should allow existing data to be exposed in whatever state it's in, without an onerous ETL project. There should be opinions about how to express units of measure, ontologies, and attribute types, but none of these should be blockers to just getting what you have today into datasets.

In summary, it's meant to have a gentle adoption curve. Easy to get started, and easy to iterate toward higher and higher levels of descriptive richness and uniformity with other data sets.

> * that the dataset is fully denormalised, so that all frequently used values for an attribute are repeated for each object?

That would depend on the data you have. This spec should accommodate datasets that are highly normalized or fully denormalized.

Looking forward, when we start talking about the "query" endpoint (which is just an empty placeholder in this PR), my hope is that we will find the cost/benefit of an existing relational query language such as SQL attractive. With its joining, filtering, grouping, and windowing capabilities, SQL can provide lots of value across that spectrum. If multiple sites agree on common schemas, the same queries can be sent to those sites in parallel (I think this could be considered a federation).

> * that multiple values for an attributed can be included?

Yes, the "schema" attribute is JSON Schema Draft 7. Each object in the "objects" array conforms to that schema. The attributes of those objects can be strings, numbers, booleans, nested objects, or arrays of those things. The entire depth of that structure can be specified by the schema; not just the root attributes.

> * that multiple values that need to have some order or precedence will capture that in their representation?

For the data itself, arrays of objects should do a good job at representing order. The schema can specify the structure of the array entries.

For the ordering of the attributes in the dataset itself, this is uncertain. I think a stable natural order for the attributes is important (imagine a CSV file that shuffles its rows every time you open it!). Although JSON doesn't specify an order among object attributes, perhaps most of today's JSON parsers preserve attribute order implicitly. If so, we can try leaving this to chance as a "SHOULD" recommendation (dataset clients SHOULD output attributes in the objects array in the same order they were specified in the schema).

> * that values for attributes of values for attributes (ad nauseam) will be captured within each object [e.g., amount for sample-volume, for [ID of aliquot-ID and date of date-of-sampling], for ID for sample-ID]?

I'm not sure I understand this properly, so forgive me if I'm answering the wrong question. :)

JSON Schema does support recursive object definitions, so you could define a tree-like structure as a dataset attribute. Another possibility (perhaps easier to work with if we go down the relational route) would be to represent each fraction of a sample as its own row, referencing a parent sample ID and a root/original sample ID. Depending on use case, this may be easier to work with than a tree of sample infos embedded in a single row.

**Comment by davidlougheed** [comment]
Oh, I also have another question [apologies, I'm fairly new to all of this] - what is / how flexible is the definition of a dataset here? For example, in the application I'm working on, researchers upload "datasets", e.g. a set of variants. This goes into a storage node which has many datasets of variants stored in a single table. For the purposes of the search API definition, is the dataset the collection of variants the researcher uploaded, or is it the entire set of variants in the database? Or option C, is it either, depending on the use case?

**Review by Relequestual — COMMENTED** [review]
JSON Schema is very permissive, allowing an empty object `{}` to be a "valid" JSON Schema (no constraints).

I wonder if really you want to impose some sort of minimum information provided in a schema that's returned. If that's the case, it's relativly easy to do for the root level, wrapping the draft-7 meta-schema with `allOf` and adding a schema which imposes some minimum required fields, such as `properties` having at least one value and having a defined `$id` and `$schema`.

Of course, you can only protect against this so far, but the intent is to SHOW intent, and suggest what the expectations are.

In terms of the OpenAPI document included here, I see no specific issues that haven't already been flagged.

**Comment by Relequestual** [comment]
SchemaBlocks is in the process of defining a service definition meta-data object.
https://github.com/ga4gh-schemablocks/playground/blob/master/sb-meta/BlockMeta.yaml

It could be useful to include this object in the OpenAPI definition file (although I'm not 100% sure that OpenAPI allows additional unknown fields... need to check).

OpenAPI 3.0.x provides a [meta schema which implies](https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v3.0/schema.yaml#L36) that unknown keywords are not allowed unless prefixed with `x-`, so you could use `x-ga4gh-doc-meta` or some such.

**Comment by antbro** [comment]
Hi Jonathan


Thanks for the answers, which are pretty much what I expected (which is a good thing!!!)


All in all it looks to me like we're going to need to accommodate relational data schema. I personally suspect that allowing full flexibility around those schema will undermine this whole approach - at least if we aim to enable larger scale federation and system integration. It will only work for limited closed networks where people agree up front the specific schema in order to federate. In such cases, an SQL approach to the querying will work of course. But that will just be using SQL on an agreed data schema - nothing really innovative there. A network could just as easily use Elasticsearch, or any other established approach.


Instead I thought the Search API project was aiming for a common way to state queries that many different projects could consume. SQL assumes too much about the data schema for that to fly, IMO. So my preference would be to concentrate on what the concept model should be, which is one step above the data schema. To then settle on an API that would be compatible with one or a few agree concept models should be doable - irrespective of how the data are stored or represented in anyone's JSON or whether they want to use SQL or ES or whatever. I cc Ian Fore, as we've just started discussing things more on the concept model level (though he's probably already in the group email).


Cheers

Tony


 Professor Anthony J Brookes
 Department of Genetics & Genome Biology
 University of Leicester
 University Road
 Leicester, LE1 7RH, UK
 Tel: +44 (0)116 2523401


________________________________

**Comment by davidlougheed** [comment]
SQL seems like a terrifying API query language to me, since it relies on people who implement the API to correctly parse/re-assemble the queries in a way which fits the underlying storage and doesn't leak any unwanted data, and the temptation with naive implementations would be to just pass it to the database directly or use half-baked sanitation measures. I agree with the thoughts mentioned here: https://github.com/ga4gh-discovery/ga4gh-case-discovery/issues/75#issuecomment-491238073

**Comment by antbro** [comment]
+1

Tony


________________________________

**Comment by jfuerth** [comment]
@davidlougheed 

> Is there going to be a way to link datasets together under some kind of common identifier/grouping system for joint-querying purposes, or is this out of scope for the search API / is the suggested way to handle this going to be to combine different types of data (ex. variants + expression levels) into a single compound dataset, at least at the API response level? [ex. querying a combined `/datasets/variants+expression_levels`]

At this point, you would join datasets by pulling them down and lining them up locally using whatever criteria make sense. Having this standard format for downloading the two datasets, storing them, and re-publishing the results is valuable because now you can write a generic tool capable of doing that join and preserving all the structural information about the joined columns along the way.

Later, as we move toward the "search" side of the spec, I feel it should be in scope to be able to express a query that asks the remote end to do the join for you. If one side of the join is much larger than the other (eg. joining variants to phenotypes), then I expect server-side joining will be much more convenient and efficient.

**Comment by jfuerth** [comment]
> Oh, I also have another question [apologies, I'm fairly new to all of this] - what is / how flexible is the definition of a dataset here? For example, in the application I'm working on, researchers upload "datasets", e.g. a set of variants. This goes into a storage node which has many datasets of variants stored in a single table. For the purposes of the search API definition, is the dataset the collection of variants the researcher uploaded, or is it the entire set of variants in the database? Or option C, is it either, depending on the use case?

My thought would be Option C.

I've been playing with an (implementation of this API that wraps Google Sheets)[https://github.com/DNAstack/ga4gh-search-adapter-google-sheets]. In that implementation, I chose to expose one dataset per worksheet.

**Comment by jfuerth** [comment]
> JSON Schema is very permissive, allowing an empty object `{}` to be a "valid" JSON Schema (no constraints).
> 
> I wonder if really you want to impose some sort of minimum information provided in a schema that's returned. If that's the case, it's relativly easy to do for the root level, wrapping the draft-7 meta-schema with `allOf` and adding a schema which imposes some minimum required fields, such as `properties` having at least one value and having a defined `$id` and `$schema`.
> 
> Of course, you can only protect against this so far, but the intent is to SHOW intent, and suggest what the expectations are.
> 
> In terms of the OpenAPI document included here, I see no specific issues that haven't already been flagged.

Thanks for this suggestion! It was definitely a goal that every dataset should at least describe its root-level attributes in the schema. There are probably realistic use cases where portions of the schema are unconstrained, so I'm comfortable with that being a possibility.

Is it common practice to create a modified copy of the Draft 7 meta-schema to introduce such constraints? There must be other specifications that build on top of JSON Schema. Is this how they do it?

**Comment by mcupak** [comment]
> SQL seems like a terrifying API query language to me, since it relies on people who implement the API to correctly parse/re-assemble the queries in a way which fits the underlying storage and doesn't leak any unwanted data, and the temptation with naive implementations would be to just pass it to the database directly or use half-baked sanitation measures. I agree with the thoughts mentioned here: [ga4gh-discovery/ga4gh-case-discovery#75 (comment)](https://github.com/ga4gh-discovery/ga4gh-case-discovery/issues/75#issuecomment-491238073)

Thanks @davidlougheed, these are important points to consider when we deal with the query language, which is beyond the scope of this PR. I filed #9 to track QL thoughts so that we don't get distracted here.

**Comment by mcupak** [comment]
> > JSON Schema is very permissive, allowing an empty object `{}` to be a "valid" JSON Schema (no constraints).
> > I wonder if really you want to impose some sort of minimum information provided in a schema that's returned. If that's the case, it's relativly easy to do for the root level, wrapping the draft-7 meta-schema with `allOf` and adding a schema which imposes some minimum required fields, such as `properties` having at least one value and having a defined `$id` and `$schema`.
> > Of course, you can only protect against this so far, but the intent is to SHOW intent, and suggest what the expectations are.
> > In terms of the OpenAPI document included here, I see no specific issues that haven't already been flagged.
> 
> Thanks for this suggestion! It was definitely a goal that every dataset should at least describe its root-level attributes in the schema. There are probably realistic use cases where portions of the schema are unconstrained, so I'm comfortable with that being a possibility.
> 
> Is it common practice to create a modified copy of the Draft 7 meta-schema to introduce such constraints? There must be other specifications that build on top of JSON Schema. Is this how they do it?

Regarding extending an external schema and what Ben described in OAS 3, for reference: https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/

**Comment by Relequestual** [comment]
> > > JSON Schema is very permissive, allowing an empty object `{}` to be a "valid" JSON Schema (no constraints).
> > > I wonder if really you want to impose some sort of minimum information provided in a schema that's returned. If that's the case, it's relativly easy to do for the root level, wrapping the draft-7 meta-schema with `allOf` and adding a schema which imposes some minimum required fields, such as `properties` having at least one value and having a defined `$id` and `$schema`.
> > > Of course, you can only protect against this so far, but the intent is to SHOW intent, and suggest what the expectations are.
> > > In terms of the OpenAPI document included here, I see no specific issues that haven't already been flagged.
> > 
> > 
> > Thanks for this suggestion! It was definitely a goal that every dataset should at least describe its root-level attributes in the schema. There are probably realistic use cases where portions of the schema are unconstrained, so I'm comfortable with that being a possibility.
> > Is it common practice to create a modified copy of the Draft 7 meta-schema to introduce such constraints? There must be other specifications that build on top of JSON Schema. Is this how they do it?
> 
> Regarding extending an external schema and what Ben described in OAS 3, for reference: https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/

Indeed. It's not modifying the JSON Scheam meta-schema, but rather think of it like (but not directly like) inheritance.

This will be easier with draft-8, but the subject matter is far bigger than you would want to discuss in a github issue! =]

**Comment by antbro** [comment]
Really basic (dumb) question....

The page title says we are discussing the "Dataset portion of the Search API". But how does this 'dataset' part contribute to the whole search concept? 

Is the idea that people with data must convert it to this format and send the JSON somewhere or feed it into some local software to be discoverable via whatever API we come up with? I don't see why that is easier than just letting the data stay at source and having a generic API query language that they can translate from into their own query engine (which is what I though Search API was about). Or is this dataset intended to play some totally different role to this?

**Comment by jfuerth** [comment]
> Really basic (dumb) question....
> 
> The page title says we are discussing the "Dataset portion of the Search API". But how does this 'dataset' part contribute to the whole search concept?
> 
> Is the idea that people with data must convert it to this format and send the JSON somewhere or feed it into some local software to be discoverable via whatever API we come up with? I don't see why that is easier than just letting the data stay at source and having a generic API query language that they can translate from into their own query engine (which is what I though Search API was about). Or is this dataset intended to play some totally different role to this?

The idea was that no matter what we end up settling on for a query language, we need a format for representing the results of a query. And this format is potentially useful beyond the "here are your query results" use case.

**Comment by jfuerth** [comment]
> All in all it looks to me like we're going to need to accommodate relational data schema. I personally suspect that allowing full flexibility around those schema will undermine this whole approach - at least if we aim to enable larger scale federation and system integration. It will only work for limited closed networks where people agree up front the specific schema in order to federate. In such cases, an SQL approach to the querying will work of course. But that will just be using SQL on an agreed data schema - nothing really innovative there.

I agree that this isn't the final destination. My argument is that we shouldn't be aiming for the final step to also be the first step. If you'll permit me a cute analogy, I think you're suggesting we need to harmonize [our data], and I'm suggesting something much more modest: we should all start by agreeing to sing something. Once we're all singing, we can come to consensus on which song to sing, choose a key, learn the lyrics and chord structure, and finally harmonize. :)

We covered the above line of thinking (without that particular analogy) on today's call; I just wanted to recap here for those who couldn't make it.

> A network could just as easily use Elasticsearch, or any other established approach. Instead I thought the Search API project was aiming for a common way to state queries that many different projects could consume. SQL assumes too much about the data schema for that to fly, IMO.

There are working examples out there in the wild of SQL frontends to ElasticSearch, so we know it's possible.

* https://www.elastic.co/blog/an-introduction-to-elasticsearch-sql-with-practical-examples-part-1
* https://github.com/NLPchina/elasticsearch-sql
* https://prestosql.io/docs/current/connector/elasticsearch.html
* https://resources.zaloni.com/blog/hive-basics-elasticsearch-integration

> So my preference would be to concentrate on what the concept model should be, which is one step above the data schema. To then settle on an API that would be compatible with one or a few agree concept models should be doable - irrespective of how the data are stored or represented in anyone's JSON or whether they want to use SQL or ES or whatever. I cc Ian Fore, as we've just started discussing things more on the concept model level (though he's probably already in the group email).

I completely agree that's got to happen before we achieve true data harmonization and powerful, frictionless federated schemas. I have given it thought. I think a dimensional model hits the sweet spot for extensibility, simplicity, and ease of (correct) use. I could share my ideas with the group on a future call, and I'd be interested in your thoughts and plans on the matter. However, I think we're two or three big steps away from implementing such plans. We're missing a lot of basic prerequisites such as a format for query results, a language for describing data values, a language for units of measure, a language for complex queries, etc.

As you said on yesterday's call, it's good to be a bit forward looking so we don't paint ourselves in to the corner, but at the same time, such long-term plans can distract us from the immediate tasks at hand.

**Comment by antbro** [comment]
I know SQL can be converted to ES etc. My point was rather that having SQL as the common query language may be flawed because "SQL assumes too much about the data schema for that to fly"

I also agree we need to tackle many levels of challenge in the whole Search API scope. But starting with the output dataset JSON spec seems a strange place to start. Also, not really a priority - as in the first instance people could just send and accept whatever file formats they wanted to, once discovered.

I prefer the idea of starting with the concept model. You mention dimensional models. Actual my group came up with exactly that, and elaborated it fully for the geno-pheno use case, as part of the previous incarnation of the Search API Team effort. The pushback we got was from people that said we did not need dimensions (or even a concept model)

**Comment by jfuerth** [comment]
As for units, this library has a BSD-like license (3-clause BSD plus an ASL2-like patent retaliation clause) and appears to be equally as capable as GNU Units: https://github.com/Unidata/UDUNITS-2/

In particular, have a look at the definitions of base, common, and compound units in the XML files in this directory: https://github.com/Unidata/UDUNITS-2/tree/master/lib

**Comment by jrambla** [comment]
Could someone explains how this aligns to other GA4GH Dataset descriptions (e.g. DRS or the the Dataset in Beacon)?
Thanks!

**Comment by jfuerth** [comment]
> Could someone explains how this aligns to other GA4GH Dataset descriptions (e.g. DRS or the the Dataset in Beacon)?
> Thanks!

This spec is intended to expose arbitrary collections of tabular data, along with unambiguous descriptions of what the attributes in each table of data mean.

The JSON format for such a table of data will be reusable for representing a Discovery Search query result: a query is expected to return zero or more items that share a structural layout. The result should be specific about what the attributes in those resulting objects represent. This matches the table format we've defined in this PR, having a JSON schema definition and an array of objects that conform to that schema.

Hope this helps!

**Review by mcupak — APPROVED** [review]

**Comment by mcupak** [comment]
Good discussion. We haven't had any activity for a week so it seems we're at a stable point where either questions have been answered without requiring further code changes, or comments diverged into the query language area, which is important to discuss, but out of scope for this PR. 

I rebased the PR so that we can merge. If you have concerns which:

- are severe enough to block the merge, please comment here now.
- can be addressed post-merge, please file an issue and comment there.
- have to do with the query language, please comment on #9.

Thanks everyone for your input!

**Pull request merged**
